beacon-chain/blockchain/service.go

```diff
@@ -63,29 +60,58 @@ func NewChainService(ctx context.Context, cfg *Config) (*ChainService, error) {
 		beaconDB:           cfg.BeaconDB,
 		web3Service:        cfg.Web3Service,
 		incomingBlockChan:  make(chan *pb.BeaconBlock, cfg.IncomingBlockBuf),
-		processedBlockChan: make(chan *pb.BeaconBlock),
+		genesisTimeChan:    make(chan time.Time),
 		incomingBlockFeed:  new(event.Feed),
 		canonicalBlockFeed: new(event.Feed),
 		canonicalStateFeed: new(event.Feed),
-		unProcessedBlocks:  make(map[uint64]*pb.BeaconBlock),
-		unfinalizedBlocks:  make(map[[32]byte]*pb.BeaconState),
 		enablePOWChain:     cfg.EnablePOWChain,
 	}, nil
 }
 
 // Start a blockchain service's main event loop.
 func (c *ChainService) Start() {
-	log.Info("Starting service")
-	var err error
-	c.genesisTime, err = c.beaconDB.GenesisTime()
+	beaconState, err := c.beaconDB.State()
 	if err != nil {
-		log.Fatalf("Unable to retrieve genesis time - blockchain service could not start: %v", err)
-		return
+		log.Fatalf("Could not attempt fetch beacon state: %v", err)
+	}
+	// If the chain has already been initialized, simply start the block processing routine.
+	if beaconState != nil {
+		log.Info("Beacon chain data already exists, starting service")
+		c.genesisTime = time.Unix(int64(beaconState.GenesisTime), 0)
+		go c.blockProcessing()
+	} else {
+		log.Info("Waiting for ChainStart log from the Validator Deposit Contract to start the beacon chain...")
+		subChainStart := c.web3Service.ChainStartFeed().Subscribe(c.genesisTimeChan)
+		go func() {
+			genesisTime := <-c.genesisTimeChan
+			log.Info("ChainStart time reached, starting the beacon chain!")
+			c.genesisTime = genesisTime
+			unixTime := uint64(genesisTime.Unix())
+			if err := c.beaconDB.InitializeState(unixTime); err != nil {
+				log.Fatalf("Could not initialize beacon state to disk: %v", err)
+			}
+			beaconState, err := c.beaconDB.State()
+			if err != nil {
+				log.Fatalf("Could not attempt fetch beacon state: %v", err)
+			}
+			// TODO(#1389): Replace by state tree hashing algorithm to determine root instead of a hash.
+			hash, err := state.Hash(beaconState)
+			if err != nil {
+				log.Fatalf("Could not hash beacon state: %v", err)
+			}
+			genesisBlock, err := c.beaconDB.BlockBySlot(0)
+			if err != nil {
+				log.Fatalf("Could not attempt to fetch genesis block: %v", err)
+			}
+			if genesisBlock == nil {
```

Comment:
we should never have this situation where `genesisBlock != nil` since the block can only be initialized after chainstart. If we face this condition, maybe an error should be logged/ or a panic and the node shutsdown?